modules/rgbd/src/tsdf_functions.cpp

```diff
@@ -373,5 +373,224 @@ void integrateVolumeUnit(
 
 }
 
+size_t calc_hash(Vec4i x)
+{
+    uint32_t seed = 0;
+    constexpr uint32_t GOLDEN_RATIO = 0x9e3779b9;
+    //uint32_t GOLDEN_RATIO = 0x9e3779b9;
+    //seed ^= x[0] + GOLDEN_RATIO + (seed << 6) + (seed >> 2);
+    //seed ^= x[1] + GOLDEN_RATIO + (seed << 6) + (seed >> 2);
+    //seed ^= x[2] + GOLDEN_RATIO + (seed << 6) + (seed >> 2);
+    for (int i = 0; i < 3; i++)
+    {
+        seed ^= x[i] + GOLDEN_RATIO + (seed << 6) + (seed >> 2);
+    }
+    return seed;
+}
+
+VolumesTable::VolumesTable()
+{
+    this->volumes = cv::Mat(hash_divisor * list_size, 1, rawType<Volume_NODE>());
+    for (int i = 0; i < volumes.size().height; i++)
+    {
+        Volume_NODE* v = volumes.ptr<Volume_NODE>(i);
+        v->idx = nan4;
+        v->row = -1;
+        v->nextVolumeRow = -1;
+        v->isActive = 0;
+        v->lastVisibleIndex = -1;
+        //v.tmp = i;
+    }
+}
+
+void VolumesTable::update(Vec3i indx)
+{
+    Vec4i idx(indx[0], indx[1], indx[2], 0);
+    int hash = int(calc_hash(idx) % hash_divisor);
+    int num  = 1;
+    int start = hash * num * list_size;
+    int i = start;
+
+    while (i != -1)
+    {
+        Volume_NODE* v = volumes.ptr<Volume_NODE>(i);
+        if (v->idx == idx)
+            return;
+        //find nan cheking for int or Vec3i
+        //if (isNaN(Point3i(v.idx)))
+        if (v->idx[0] == -2147483647)
+        {
+            v->idx = idx;
+            v->nextVolumeRow = getNextVolume(hash, num, i, start);
+            indexes.push_back(indx);
+            indexesGPU.push_back(idx);
+            return;
+        }
+        i = v->nextVolumeRow;
+    }
+}
+
+void VolumesTable::update(Vec3i indx, int row)
+{
+    Vec4i idx(indx[0], indx[1], indx[2], 0);
+    int hash = int(calc_hash(idx) % hash_divisor);
+    int num = 1;
+    int start = hash * num * list_size;
```

Comment:
To see that this line is incorrect, just imagine that `hash` is zero while `num` is not.
They would have the same `start` value for all possible `num`.
Please fix the formula.

Reply:
done